libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,646 @@
+        __it_.template emplace<_Idx + 1>(ranges::begin(std::get<_Idx + 1>(__parent_->__views_)));
+        __advance_fwd<_Idx + 1>(0, __offset + __steps - __n_size);
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __advance_bwd(difference_type __offset, difference_type __steps) {
+    using __underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, __base_iter>>;
+    if constexpr (_Idx == 0) {
+      std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+    } else {
+      if (__offset >= __steps) {
+        std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+      } else {
+        auto __prev_size = ranges::__distance(std::get<_Idx - 1>(__parent_->__views_));
+        __it_.template emplace<_Idx - 1>(ranges::begin(std::get<_Idx - 1>(__parent_->__views_)) + __prev_size);
+        __advance_bwd<_Idx - 1>(__prev_size, __steps - __offset);
+      }
+    }
+  }
+
+  template <size_t... _Is, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func, index_sequence<_Is...>) {
+    ((__index == _Is ? (__func(integral_constant<size_t, _Is>{}), 0) : 0), ...);
+  }
+
+  template <size_t _Idx, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func) {
+    __apply_at_index(__index, std::forward<_Func>(__func), make_index_sequence<_Idx>{});
+  }
+
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI explicit constexpr __iterator(__maybe_const<_Const, concat_view>* __parent, _Args&&... __args)
+    requires constructible_from<__base_iter, _Args&&...>
+      : __it_(std::forward<_Args>(__args)...), __parent_(__parent) {}
+
+public:
+  _LIBCPP_HIDE_FROM_ABI __iterator() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator(__iterator<!_Const> __i)
+    requires _Const && (convertible_to<iterator_t<_Views>, iterator_t<const _Views>> && ...)
+      : __it_([&__src = __i.it_]<size_t... _Indices>(size_t __idx, index_sequence<_Indices...>) -> __base_iter {
+          using __src_lref          = decltype((__src));
+          using __construction_fptr = __base_iter (*)(__src_lref);
+          static constexpr __construction_fptr __vtable[]{[](__src_lref __src_var) -> __base_iter {
+            return __base_iter(in_place_index<_Indices>, std::get<_Indices>(std::move(__src_var)));
+          }...};
+          return __vtable[__idx](__src);
+        }(__i.index(), make_index_sequence<variant_size_v<__base_iter>>{})),
+        __parent_(__i.parent_) {}
+
+  _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator*() const {
+    return std::visit(
+        [](auto&& __it) -> __concat_reference_t<__maybe_const<_Const, _Views>...> { return *__it; }, __it_);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator& operator++() {
+    size_t __active_index = __it_.index();
+    __apply_at_index<std::variant_size_v<decltype(__it_)>>(__active_index, [&](auto __index_constant) {
```

Comment:
```suggestion
    __apply_at_index<variant_size_v<decltype(__it_)>>(__active_index, [&](auto __index_constant) {
```

Ditto for other `std::variant_size_v`.